### PR TITLE
[New Task] Function App Deploy - Windows / Linux

### PR DESCRIPTION
**Pull request by SumiranAgg**

### CODE REVIEW

File: Tasks/AzureFunctionDeploymentV1/README.md [hunk 1]
```diff
@@ -0,0 +1,117 @@
+#  Azure Function Deployment: ARM 
+
+## Overview
+
+The Azure Function Deployment task is used to update Azure App Service to deploy [Functions](https://docs.microsoft.com/en-us/azure/azure-functions/) to Azure. The task works on cross platform Azure Pipelines agents running Windows, Linux or Mac and uses the underlying deployment technologies of RunFromPackage, Zip Deploy and [Kudu REST APIs](https://github.com/projectkudu/kudu/wiki/REST-API).
+
+The task works for [ASP.NET](https://www.visualstudio.com/en-us/docs/release/examples/azure/azure-web-apps-from-build-and-release-hubs), [ASP.NET Core](https://www.visualstudio.com/en-us/docs/release/examples/azure/aspnet-core10-azure-web-apps), PHP, Java, Python, Go and [Node.js](https://www.visualstudio.com/en-us/docs/release/examples/nodejs/node-to-azure-webapps) based web applications.
```
Comment (RoshanKumarMicrosoft) [inline Tasks/AzureFunctionDeploymentV1/README.md]: For each url it is redirecting to same landing page. Is this expected ?

File: Tasks/AzureFunctionDeploymentV1/README.md [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+* **Select deployment method:** Select the option to to choose from Zip Deploy, RunFromPackage
+
+By default the task tries to select the appropriate deployment technology given the input package, app service type and agent OS.
+
+* **Generate Web.config:** A standard Web.config will be generated and deployed to Azure App Service if the application does not have one. For example, for [Nodejs application, web.config](https://github.com/projectkudu/kudu/wiki/Using-a-custom-web.config-for-Node-apps) will have startup file and iis_node module values. Similarly for Python (Bottle, Django, Flask) the web.config will have details of WSGI handler, Python path etc. The task will generate a new web.config only when the artifact package/folder does not contain an existing web.config. The default values populated by the task can be overriden in the task by using the Web.config parameters field.
```
Comment (RoshanKumarMicrosoft) [inline Tasks/AzureFunctionDeploymentV1/README.md]: Can you please also specify for jar it is also generating web.config?

File: Tasks/AzureFunctionDeploymentV1/README.md [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+* **Web.config parameters:** Edit values like startup file in the task generated web.config file. The default values populated by the task can be overridden in the task by passing the web.config parameters. This edit feature is **only for the generated web.config**. Feature is useful when [Azure App Service Manage task](https://github.com/Microsoft/vsts-tasks/tree/master/Tasks/AzureAppServiceManage) is used to install specific Python version by using extensions or when you want to provide a different startup file for Node.js. 
+In  case of Python, the path can be set as an output variable of the [Azure App Service Manage task](https://github.com/Microsoft/vsts-tasks/tree/master/Tasks/AzureAppServiceManage)  and then set as the Python path in the web.config generated by this deploy task. You can try out this feature by selecting any Python, Nodejs, PHP release definition template.
```
Comment (RoshanKumarMicrosoft) [inline Tasks/AzureFunctionDeploymentV1/README.md]: Java is missing in list. Please also add that.

File: Tasks/AzureFunctionDeploymentV1/deploymentProvider/BuiltInLinuxWebAppDeploymentProvider.ts [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+const linuxFunctionStorageSetting: string = '-WEBSITES_ENABLE_APP_SERVICE_STORAGE true';
+const linuxFunctionRuntimeSettingName: string = '-FUNCTIONS_WORKER_RUNTIME ';
+
+const linuxFunctionRuntimeSettingValue = new Map([
+    [ 'DOCKER|microsoft/azure-functions-dotnet-core2.0:2.0', 'dotnet ' ],
+    [ 'DOCKER|microsoft/azure-functions-node8:2.0', 'node ' ]
+]);
+
+export class BuiltInLinuxWebAppDeploymentProvider extends AzureRmWebAppDeploymentProvider{
+    private zipDeploymentID: string;
```
Comment (RoshanKumarMicrosoft) [inline Tasks/AzureFunctionDeploymentV1/deploymentProvider/BuiltInLinuxWebAppDeploymentProvider.ts]: Space between name and {

File: Tasks/AzureFunctionDeploymentV1/task.json [hunk 5]
```diff
@@ -0,0 +1,381 @@
+{
+    "id": "501DD25D-1785-43E4-B4E5-A5C78CCC0573",
+    "name": "AzureFunctionDeployment",
+    "friendlyName": "Azure Function",
+    "description": "Update Azure Function on Windows, Function on Linux with built-in images, ASP.NET, .NET Core, PHP, Python or Node.js based Web applications",
+    "helpMarkDown": "[More information](https://aka.ms/azurefunctiondeployreadme)",
```
Comment (RoshanKumarMicrosoft) [inline Tasks/AzureFunctionDeploymentV1/task.json]: Broken Url

File: Tasks/AzureFunctionDeploymentV1/task.json [hunk 6]
```diff
@@ -0,0 +1,381 @@
...
+        "DeploymentGroup"
+    ],
+    "author": "Microsoft Corporation",
+    "version": {
+        "Major": 1,
+        "Minor": 0,
+        "Patch": 0
+    },
+    "releaseNotes": "What's new in version 1.* (preview)<br />Supports Zip Deploy, Run From Package, War Deploy<br />Supports App Service Environments<br />Improved UI for discovering different App service types supported by the task<br/>Run From Package is the preferred deployment method, which makes files in wwwroot folder read-only<br/>Click [here](https://aka.ms/azurefunctiondeployreadme) for more information.",
+    "minimumAgentVersion": "2.104.1",
```
Comment (RoshanKumarMicrosoft) [inline Tasks/AzureFunctionDeploymentV1/task.json]: Is there any reason not supporting 1.x version of agent ?

File: Tasks/AzureFunctionDeploymentV1/taskparameters.ts [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+export enum DeploymentType {
+    auto,
+    zipDeploy,
+    runFromPackage,
+    warDeploy
+}
+
+export interface TaskParameters {
+    connectedServiceName: string;
+    WebAppName?: string;
```
Comment (RoshanKumarMicrosoft) [inline Tasks/AzureFunctionDeploymentV1/taskparameters.ts]: Is webAppName can be null ? tl.getInput('appName', true); here it is mandatory field.

File: Tasks/AzureFunctionDeploymentV1/deploymentProvider/WindowsWebAppZipDeployProvider.ts [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+        if(tl.stats(webPackage).isDirectory()) {
+            let tempPackagePath = deployUtility.generateTemporaryFolderOrZipPath(tl.getVariable('AGENT.TEMPDIRECTORY'), false);
+            webPackage = await zipUtility.archiveFolder(webPackage, "", tempPackagePath);
+            tl.debug("Compressed folder into zip " +  webPackage);
+        }
+
+        tl.debug("Initiated deployment via kudu service for webapp package : ");
+        
+        var updateApplicationSetting = ParameterParser.parse(removeRunFromZipAppSetting)
```
Comment (RoshanKumarMicrosoft) [inline Tasks/AzureFunctionDeploymentV1/deploymentProvider/WindowsWebAppZipDeployProvider.ts]: var updateApplicationSetting = ParameterParser.parse(removeRunFromZipAppSetting)
  var deleteApplicationSetting = ParameterParser.parse(deleteOldRunFromZipAppSetting)

Can we put this inside appServiceUtility (updateAndMonitorAppSettings) then all provider can reuse it and code will looks clear .

File: Tasks/AzureFunctionDeploymentV1/deploymentProvider/WindowsWebAppZipDeployProvider.ts [hunk 9]
```diff
@@ -0,0 +1,58 @@
+import { AzureRmWebAppDeploymentProvider } from './AzureRmWebAppDeploymentProvider';
+import tl = require('vsts-task-lib/task');
+import * as ParameterParser from 'azurermdeploycommon/operations/ParameterParserUtility'
+import { DeploymentType } from '../taskparameters';
+import { PackageType } from 'azurermdeploycommon/webdeployment-common/packageUtility';
+import { FileTransformsUtility } from 'azurermdeploycommon/operations/FileTransformsUtility.js';
+const deleteOldRunFromZipAppSetting: string = '-WEBSITE_RUN_FROM_ZIP';
```
Comment (RoshanKumarMicrosoft) [inline Tasks/AzureFunctionDeploymentV1/deploymentProvider/WindowsWebAppZipDeployProvider.ts]: Please put inside appServiceUtility

File: Tasks/AzureFunctionDeploymentV1/deploymentProvider/WindowsWebAppRunFromZipProvider.ts [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+
+        if(tl.stats(webPackage).isDirectory()) {
+            let tempPackagePath = deployUtility.generateTemporaryFolderOrZipPath(tl.getVariable('AGENT.TEMPDIRECTORY'), false);
+            webPackage = await zipUtility.archiveFolder(webPackage, "", tempPackagePath);
+            tl.debug("Compressed folder into zip " +  webPackage);
+        }
+
+        tl.debug("Initiated deployment via kudu service for webapp package : ");
+        
+        var addCustomApplicationSetting = ParameterParser.parse(runFromZipAppSetting);
```
Comment (RoshanKumarMicrosoft) [inline Tasks/AzureFunctionDeploymentV1/deploymentProvider/WindowsWebAppRunFromZipProvider.ts]: Please put inside appServiceUtility.

File: Tasks/AzureFunctionDeploymentV1/deploymentProvider/BuiltInLinuxWebAppDeploymentProvider.ts [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+        }
+        
+        switch(this.taskParams.Package.getPackageType()){
+            case PackageType.folder:
+                let tempPackagePath = webCommonUtility.generateTemporaryFolderOrZipPath(tl.getVariable('AGENT.TEMPDIRECTORY'), false);
+                let archivedWebPackage = await zipUtility.archiveFolder(this.taskParams.Package.getPath(), "", tempPackagePath);
+                tl.debug("Compressed folder into zip " +  archivedWebPackage);
+                this.zipDeploymentID = await this.kuduServiceUtility.deployUsingZipDeploy(archivedWebPackage);
+            break;
+            case PackageType.zip:
```
Comment (RoshanKumarMicrosoft) [inline Tasks/AzureFunctionDeploymentV1/deploymentProvider/BuiltInLinuxWebAppDeploymentProvider.ts]: Please put Space between zip and break .

**Comment by RoshanKumarMicrosoft** [comment]
Please add link of **java function app deployment through our task** So it will be very helpful for user .

**Comment by RoshanKumarMicrosoft** [comment]
Please add L0 for this.

File: Tasks/AzureFunctionDeploymentV1/README.md (hunk 1 above)
Comment (SumiranAgg) [reply Tasks/AzureFunctionDeploymentV1/README.md]: Yes. This is same as parent task. Will cross check if has to be replaced.

File: Tasks/AzureFunctionDeploymentV1/README.md (hunk 2 above)
Comment (SumiranAgg) [reply Tasks/AzureFunctionDeploymentV1/README.md]: have to get this reviewed with PMs. Will update accordingly.

File: Tasks/AzureFunctionDeploymentV1/README.md (hunk 3 above)
Comment (SumiranAgg) [reply Tasks/AzureFunctionDeploymentV1/README.md]: have to get this reviewed with PMs. Will update accordingly.

**Pull request merged**